Add IsMinimumDocument method to WordprocessingDocument, SpreadsheetDocument, and PresentationDocument

mikeebowen opened the pull request:
- Adds methods to determine if a document is a minimum document for WordprocessingDocument, SpreadsheetDocument, and PresentationDocument.
- closes #816

twsouthwick reviewed (changes requested):
This is a great idea. However, it would be great for it to have the following characteristics:

- Be overrideable/replaceable
- Be accessed off of OpenXmlPackage directly rather than implemented separately

Questions that would decide the API shape:

- Should this be a separate API? Or should it be opt in while opening a package? I tend not to be a fan of APIs that check if something is valid then doing it - there's always the potential something gets changed underneath
- Do we really want to validate the extension? What about if it's a stream/package? If it is overrideable, we could have a check if we've opened it with a path to validate that if needed
- Are we OK with it throwing an exception if it is an ill-formed package (i.e. we already do that)?

mikeebowen commented:
@twsouthwick 

> Be overrideable/replaceable

Created virtual method in `OpenXmlPackage`

> Be accessed off of OpenXmlPackage directly rather than implemented separately

There is a default implementation in `OpenXmlPackage` but the individual Wordprocessing, Spreadsheet, and Presentation Documents have unique implementations for their different requirements.

> Should this be a separate API? Or should it be opt in while opening a package? I tend not to be a fan of APIs that check if something is valid then doing it - there's always the potential something gets changed underneath

Added `VerifyMinimumPackage` option to `OpenSettings` when true, the package contents are examined and throw an exception if minimum package requirements are not met

> Do we really want to validate the extension? What about if it's a stream/package? If it is overrideable, we could have a check if we've opened it with a path to validate that if needed

The minimum package is verified on all Open overloads with `VerifyMinimumPackage` true. The Open method handles exceptions for the file path such as invalid extension or invalid characters

> Are we OK with it throwing an exception if it is an ill-formed package (i.e. we already do that)?

Removed file checks. The Open method already handles those.

mikeebowen commented on src/DocumentFormat.OpenXml/Packaging/WordprocessingDocument.cs:
```diff
@@ -310,6 +310,96 @@ public static WordprocessingDocument Open(Package package, OpenSettings openSett
+            catch
+            {
+                return false;
+            }
```
No longer applicable

mikeebowen commented on src/DocumentFormat.OpenXml/Packaging/SpreadsheetDocument.cs:
```diff
@@ -267,6 +268,111 @@ public static SpreadsheetDocument Open(System.IO.Stream stream, bool isEditable)
+            catch
+            {
+                return false;
+            }
```
No longer applicable

mikeebowen commented on src/DocumentFormat.OpenXml/Packaging/PresentationDocument.cs:
```diff
@@ -267,6 +270,120 @@ public static PresentationDocument Open(Package package, OpenSettings openSettin
+            catch
+            {
+                return false;
+            }
```
No longer applicable

twsouthwick commented:
@mikeebowen this would be a good use case for a feature so it can be overridden at runtime vs compile time. Here's an example of how to do it: https://github.com/dotnet/Open-XML-SDK/commit/fa893dd6686124d2490ba92f7178b64308448698

twsouthwick commented:
A few questions about the direction/goal here

**Only a few kinds of packages are supported (such as in the presentation doc) - it will fail if it's an unsupported one**

Is this what we want? I'd expect it to validate it but fail if it is invalid. It appears to also fail if we don't have a check which seems a little unexpected.

**Validate() throws and returns true/false**

I'd expect it to do one or the other. Is the manual throwing due to custom messages? If so, let's make it consistent

**Mixture of exception types**

I think we should use an existing OpenXmlPackageException type or create a new one for this

**How to think about this vs validation**

At some point, how is this different from the existing OpenXmlValidator logic? Should we be hooking this up to that and performing a full validation?

twsouthwick commented on src/DocumentFormat.OpenXml.Framework/Validation/DocumentValidator.cs:
```diff
@@ -106,6 +107,20 @@ private void ValidatePart(OpenXmlPart part, ValidationContext context)
                     {
                         Validate(context);
                     }
+                    else if (part.Uri.ToString().EndsWith(".xml", System.StringComparison.InvariantCultureIgnoreCase) &&
```
need to dispose the stream from getpart. probably best to add an extension method of ".IsEmptyPart()" that can handle correctly disposing the stream

mikeebowen replied:
Done

twsouthwick commented on src/DocumentFormat.OpenXml.Framework/Validation/DocumentValidator.cs:
```diff
@@ -106,6 +107,19 @@ private void ValidatePart(OpenXmlPart part, ValidationContext context)
                     {
                         Validate(context);
                     }
+                    else if (part.Uri.ToString().EndsWith(".xml", System.StringComparison.InvariantCultureIgnoreCase) && part.IsEmptyPart())
```
this check should be moved into the IsEmptyPart method

twsouthwick commented on src/DocumentFormat.OpenXml.Framework/Packaging/OpenXmlPart.cs:
```diff
@@ -448,6 +448,18 @@ internal virtual bool IsInVersion(FileFormatVersions version)
 
+        internal override bool IsEmptyPart()
+        {
+            bool isEmptyPart;
+
+            using (Stream stream = GetStream())
+            {
+                isEmptyPart = stream.Length == 0;
+            }
+
+            return isEmptyPart;
```
```suggestion
            using (Stream stream = GetStream())
            {
                return stream.Length == 0;
            }
```
